src/Text/Layout/Table.hs

```diff
@@ -416,21 +416,26 @@ tableLinesBWithCMIs specs TableStyle { .. } rowHeader colHeader rowGroups =
             , bothHeadersT
             )
 
+    emptyFromCMI = spacesB . widthCMI
+
     cMSs      = map cutMark specs
     posSpecs  = map position specs
-    cMIs      = fitHeaderIntoCMIs $ deriveColModInfos' specs $ concatMap rows rowGroups
-    rowMods   = zipWith3 columnModifier posSpecs cMSs cMIs
-    rowBody   = map (zipWith ($) rowMods) . rows
+    cMIs      = fitHeaderIntoCMIs $ deriveColModInfosFromColumns' specs $ transposeRowGroups rowGroups
+    rowMods   = zipWith3 (\p cm cmi -> (emptyFromCMI cmi, columnModifier p cm cmi)) posSpecs cMSs cMIs
+
+    rowBody :: RowGroup a -> [[b]]
+    rowBody   = mapRowGroupColumns rowMods
     colWidths = map widthCMI cMIs
 
     -- Apply modifiers to rows, adding row headers to the first row in the group if needed
     applyRowMods  :: (Maybe (HeaderColSpec, r), RowGroup a) -> [(Maybe b, [b])]
     applyRowMods (Just (hSpec, r), grp) | Just rCMI <- rowHeaderCMI
         = zip (header rCMI) (rowBody grp)
       where
-        header cMI = map (Just . headerCellModifier hSpec noCutMark cMI) $ r : repeat emptyCell
+        header cMI = (Just $ headerCellModifier hSpec noCutMark cMI r) : repeat (Just $ emptyFromCMI cMI)
```

Comment:
Optional: replace this with `map Just $ headerCellModifier hSpec noCutMark cMI r : repeat (emptyFromCMI cMI)`. It makes it a bit clearer that everything's guaranteed to be `Just`.